packages/vector_graphics_compiler/lib/src/svg/masking_optimizer.dart

```diff
@@ -0,0 +1,367 @@
+          ));
+        }
+        break;
+      case path_ops.PathVerb.close:
+        newCommands.add(const CloseCommand());
+        break;
+    }
+  }
+
+  Path newPath = Path(commands: newCommands);
+
+  if (path.fillType == path_ops.FillType.evenOdd) {
+    newPath = Path(commands: newCommands, fillType: PathFillType.evenOdd);
+  }
+
+  return newPath;
+}
+
+/// Gets the single child recursively,
+/// returns null if there are 0 children or more than 1
+ResolvedPathNode? getSingleChild(Node node) {
+  if (node.runtimeType.toString() == 'ResolvedPathNode') {
+    return node as ResolvedPathNode;
+  } else if (node.runtimeType.toString() == 'ParentNode') {
+    if ((node as ParentNode).children.length == 1) {
+      return getSingleChild(node.children.single);
+    } else {
+      return null;
+    }
+  } else {
+    return null;
+  }
+}
+
+/// Traverses all children of a given node
+List<T> queryChildren<T extends Node>(Node node) {
+  final List<T> children = <T>[];
+  void visitor(Node child) {
+    if (child is T) {
+      children.add(child);
+    }
+    child.visitChildren(visitor);
+  }
+
+  node.visitChildren(visitor);
+  return children;
+}
+
+/// Checks if one shape is inside another using rectagle bounds
+bool isInside(Rect outerShape, Rect innerShape) {
+  return ((innerShape.left > outerShape.left) &&
+      (innerShape.right < outerShape.right) &&
+      (innerShape.top > outerShape.top) &&
+      (innerShape.bottom < outerShape.bottom));
+}
+
+/// Parses SVG string to node tree and runs it through the resolver
+Future<Node> parseAndResolve(String source) async {
+  final Node node = await parseToNodeTree(source);
+  final ResolvingVisitor visitor = ResolvingVisitor();
+  return node.accept(visitor, AffineMatrix.identity);
+}
+
+/// Simplifies masking operations into PathNodes
+class MaskingOptimizer extends Visitor<_Result, Node>
+    with ErrorOnUnResolvedNode<_Result, Node> {
+  ///List of masks to add.
+  List<ResolvedPathNode> masksToApply = [];
+
+  /// Applies visitor to given node
+  Node apply(Node node) {
+    final Node newNode = node.accept(this, null).node;
+    return newNode;
+  }
+
+  /// Applies mask to a path node, and returns resulting path node.
+  ResolvedPathNode applyMask(Node child, ResolvedPathNode maskPathNode) {
+    final ResolvedPathNode pathNode = child as ResolvedPathNode;
+    final path_ops.Path maskPathOpsPath = toPathOpsPath(maskPathNode.path);
+    maskPathOpsPath.applyOp(
+        toPathOpsPath(pathNode.path), path_ops.PathOp.intersect);
+    final Path newPath = toVectorGraphicsPath(maskPathOpsPath);
+    final ResolvedPathNode newPathNode = ResolvedPathNode(
+        paint: pathNode.paint, bounds: maskPathNode.bounds, path: newPath);
+    return newPathNode;
+  }
+
+  @override
+  _Result visitEmptyNode(Node node, void data) {
+    final _Result _result = _Result(node);
+    return _result;
+  }
+
+  @override
+  _Result visitChildren(Node node, void data) {
+    throw UnimplementedError();
```

Comment:
This makes me think that the `data` param should be a forward passed `_Result` and just get returned here...

Reply:
fixed.

Reply:
I'm a little confused by what you mean in this?

Reply:
I think this method should be implemented, but to implement it you'll probably need something from `data` to build the `_Result` you want to return.

Reply:
e.g. I think a signature like 

```dart
_Result visitChildren(Node node, _Result data) {
  return data;
}
```

Might make sense. But I might be missing something.

Reply:
oh ok, that makes sense.